### PR TITLE
Feature/1050 options unique

### PR DESCRIPTION
**Pull request by kzangeli**
Fixes issue #1050

### CODE REVIEW

File: src/lib/ngsi/ContextAttributeVector.h [hunk 1]
```diff
@@ -65,10 +65,14 @@ typedef struct ContextAttributeVector
...
-                            bool                omitValue = false,
+                            bool                comma       = false,
+                            bool                omitValue   = false,
                             bool                attrsAsName = false);
-  std::string        toJson(bool isLastElement, bool types, const std::string& renderMode, const std::string& attrsFilter = "");
+
+  std::string        toJson(bool                isLastElement,
+                            bool                types,
+                            const std::string&  renderMode,
+                            const std::string& attrsFilter  = "");
```
Comment (fgalan) [inline src/lib/ngsi/ContextAttributeVector.h]: Align (one whitespace missing just before `attrsFilter`).

File: src/lib/rest/ConnectionInfo.cpp [hunk 2]
```diff
@@ -91,6 +92,13 @@ int uriParamOptionsParse(ConnectionInfo* ciP, const char* value)
     ciP->uriParamOptions[vec[ix]] = true;
   }
 
+  //
+  // Check of invalid combinations
+  //
+  if (ciP->uriParamOptions["keyValues"] && ciP->uriParamOptions["values"])   return -1;
```
Comment (fgalan) [inline src/lib/rest/ConnectionInfo.cpp]: We are using the same "keyValues", "values", etc. in several parts. Using a #define for them will help to make the code easier to mantain.

File: test/functionalTest/cases/1050_options_unique/options_unique.test [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+Date: REGEX(.*)
+
+[
+    "Mercedes",
+    "BMW",
+    "Audi"
+]
+
+
+04. URI param options with both 'keyValues' and 'values', see error
```
Comment (fgalan) [inline test/functionalTest/cases/1050_options_unique/options_unique.test]: Maybe the "conflicting options" steps should be moved to a separate .test.

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 4]
```diff
@@ -827,6 +831,8 @@ Response:
           corresponding section for details.
           + values - when used, the response payload uses `values` simplified entity representation. See corresponding 
           section for details.
+          + values - when used, the response payload uses `values` simplified entity representation. And, recurring values are left out.
```
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Copy-paste typo: it shouldn't be "values"

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 5]
```diff
@@ -1690,6 +1696,8 @@ Response code:
           section for details.
           + values - when used, the response payload uses `values` simplified entity representation. See corresponding
           section for details.
+          + unique - when used, the response payload uses `values` simplified entity representation. And, recurring values are left out.
+          See corresponding section for details.
```
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: greping in current .apib shows 4 occurrences of "+ values". However, this PR only includes three additions (typos appart) from "+ unique". Maybe one piece of documentation is missing yet?

File: CHANGES_NEXT_RELEASE [hunk 6]
```diff
@@ -3,3 +3,4 @@
 - Fix: check for metadatas with empty name (Issue #1438)
 - Fix: accepting backslash-slash in JSON v1 payload (Issue #1852)
 - Remove: deprecated functionality related with ONTIMEINTERVAL subscriptions
+- Add: new value for URI param 'options': unique - like 'values', but avoiding duplicates (Issue #1050)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: I understand that "unique" vs. "distinct" issue is still opened...

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 5 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: I have just realized that the missing operation is `GET /v2/types`, in which case unique/distinct doesn't make sense (as entity types returned by that operation are different by definition). Thus, everything seems to be ok (except the typo mentioned in other comment).

NTC

Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Ah yes, I actually saw that when searching the file for 'values' ...
Sorry for not responding. I kind of forgot.

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 4 above)
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 54a4c84

File: test/functionalTest/cases/1050_options_unique/options_unique.test (hunk 3 above)
Comment (kzangeli) [inline test/functionalTest/cases/1050_options_unique/options_unique.test]: Yeah, they could. However, the test file was so small any way ...

If you insist, I have no problem in moving the 'error part' to another functest ;-)

File: CHANGES_NEXT_RELEASE (hunk 6 above)
Comment (kzangeli) [inline CHANGES_NEXT_RELEASE]: Yes, this is not 100% decided yet.
Very easy to change, if necessary.

File: test/functionalTest/cases/1050_options_unique/options_unique.test (hunk 3 above)
Comment (fgalan) [inline test/functionalTest/cases/1050_options_unique/options_unique.test]: I think it is not a matter of size, but a matter of having unrelated topic separated (easier to mantain if we need to modify the test). The test for checking "rendering options" mutual exclusion could be even in a different case subdirectory.

Comment (kzangeli) [inline test/functionalTest/cases/1050_options_unique/options_unique.test]: Sure.

Fixed in fca6feb

File: src/lib/ngsi/ContextAttributeVector.h (hunk 1 above)
Comment (kzangeli) [inline src/lib/ngsi/ContextAttributeVector.h]: Fixed in 3abbb27

File: src/lib/rest/ConnectionInfo.cpp (hunk 2 above)
Comment (kzangeli) [inline src/lib/rest/ConnectionInfo.cpp]: That and a little more fixed in 3abbb27

File: CHANGES_NEXT_RELEASE (hunk 6 above)
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: NTC at the end.

File: src/lib/common/globals.h [hunk 7]
```diff
@@ -82,6 +85,34 @@
...
+#define RENDER_MODE_UNIQUE_VALUES "unique"
+
+
+
+/* ****************************************************************************
+*
+* Values for the URI param 'options'
+*/
+#define OPT_COUNT           "count"
+#define OPT_TEXT            "text"
```
Comment (fgalan) [inline src/lib/common/globals.h]: "text" is no longer an options... Maybe could you use this PR for the little change of removing it?

Comment (kzangeli) [inline src/lib/common/globals.h]: Sure, no probs

Comment (kzangeli) [inline src/lib/common/globals.h]: Fixed in fb486d4

**Comment by fgalan** [comment]
LGTM (passing the ball to @crbrox for final LGTM + merge + delete branch).

Note that sync from develop is needed to make this PR mergeable.

**Comment by crbrox** [comment]
LGTM

**Pull request merged**